### PR TITLE
Added Burned in Annotation Warnings

### LINKED ISSUE #44: Consider warning if Burned In Annotation is YES
If [Burned In Annation](https://dicom.innolitics.com/ciods/cr-image/general-image/00280301) is "YES", our anonymization won't be sufficient.

We could emit a WARN level log message to indicate that there are images with burned-in demographics.

If the attribute is absent, our anonymization still might not be sufficient, so we have to decide what to do about that. I suspect it's often absent, so we might want to take no action.

We could add a flag so users can customize the behaviour. For example
* never warn
* warn if Burned In Annotation is "YES"
* warn unless Burned In Annotation is "NO"

### PR DESCRIPTION
**Pull request by justineclin**
Fixes #44.

### CODE REVIEW

**Comment by justineclin** [comment]
Not sure why tests/orig_data/test_burned_in_warning_logged_at_warning/p01_s01_s01_i01.dcm says "Empty file" in the commit diff. Perhaps .dcm files aren't compared? That file isn't empty on my local folder and seems to have been committed. I think it's just not showing in this PR and the commit went in okay.

**Review by blairconrad — COMMENTED** [review]
Thanks, @justineclin. Generally I like it. I especially appreciate the comprehensive tests.
I made two minor line-level comments, which would not prevent me from merging, but there was something else that caught my eye as I read, and I confirmed with this test:

```
❯ py -m dicognito --burned-in-annotation-warning unless-no ..\tests\orig_data\test_directory_is_recursed\
Burned In Annotation attribute value does not exist.
Burned In Annotation attribute value does not exist.
```

If I'm anonymizing multiple files, and one or more have violations, the output doesn't tell me which file(s) are the odd ones out.
Your addition is better than no warning, but still may leave me scrambling to figure out where my problem lies.

**Comment by blairconrad** [comment]
> Not sure why tests/orig_data/test_burned_in_warning_logged_at_warning/p01_s01_s01_i01.dcm says "Empty file" in the commit diff. Perhaps .dcm files aren't compared? 

Generally a diff isn't shown for binary files, but there's an indication that "binary files differ" or some such.
The display looks weird to me. But the file looks good in my sandbox, as you say, and if you ask nice GitHub will at least tell you its size:
https://github.com/blairconrad/dicognito/blob/3297b569ffe6c9fb7d7ce702f50087442ad783cd/tests/orig_data/test_burned_in_warning_logged_at_warning/p01_s01_s01_i01.dcm
I wouldn't worry about it.

**Comment by justineclin** [comment]
Thanks so much for your comments, @blairconrad . Glad you enjoyed the tests. You'd be proud of me - I wrote the tests first before the code, made sure they failed, then fixed them with the implementation (superman pose) </brag>

Oh shoot, you're right! I will look to handle the multi file case and come back! (And address the line comments).

Thanks again - enjoy your weekend!

File: src/dicognito/burnedinannotationwarner.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        -------
+        The warning message.
+        """
+        if "BurnedInAnnotation" not in dataset:
+            if self.warn_level == "unless-no":
+                return "Burned In Annotation attribute value does not exist."
+        elif (self.warn_level == "if-yes" and dataset.BurnedInAnnotation == "YES") or (
+            self.warn_level == "unless-no" and dataset.BurnedInAnnotation != "NO"
+        ):
+            return "Burned In Annotation attribute value is {}.".format(dataset.BurnedInAnnotation)
```
[the remark this answers is not in the document]
Comment (justineclin) [reply src/dicognito/burnedinannotationwarner.py:41]: Oh! f-strings are awesome - will change this - thanks so much for making it known!

**Comment by justineclin** [comment]
May I get your take on something, @blairconrad?

To address your comment about anonymizing multiple studies, my thinking is to identify each study using filename for which we need to report a burned in annotation status, like this:

`PS D:\GitRepos\dicognito-J\src> py -m dicognito --burned-in-annotation-warning unless-no --output-directory d:\tmp\testDicog ..\tests\orig_data\test_directory_is_recursed\
File ..\tests\orig_data\test_directory_is_recursed\p01_s01_s01_i01.dcm: Burned In Annotation attribute value does not exist.
File ..\tests\orig_data\test_directory_is_recursed\a\b\p01_s02_s01_i01.dcm: Burned In Annotation attribute value does not exist.
Accession Number Patient ID       Patient Name
---------------- ----------       ------------
8ZKXQMGW8FIC     1E5RVWVBG916     GALLEGOS^CLEO^SUSANNE
S3UJF7QIMZVW     1E5RVWVBG916     GALLEGOS^CLEO^SUSANNE`

Thoughts? Or you think Study UID makes better sense?

**Comment by blairconrad** [comment]
Hey. I think filename makes more sense. Why make the user crack open their input files looking for the matching SOP Instance UID?

**Comment by justineclin** [comment]
I thought so too - just thought I'd ask in case I missed something.
Great - I'll do that.

Thanks for the input!
- Justin

File: src/dicognito/burnedinannotationwarner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (justineclin) [reply src/dicognito/burnedinannotationwarner.py:41]: When I Googled, there's another style of f-String that's shown. Looks like it got introduced in Python 3?
https://www.datacamp.com/community/tutorials/f-string-formatting-in-python

Not sure if this is what you have in mind. I'll upload the changes. Let me know what you think. Also happy to scan the files and refactor my code or the rest if we'd like to keep the formatting consistent.

**Comment by justineclin** [comment]
Also, I'm not sure when the squashing happens. I kept the commits to make the additional addressing of review comments clear. I can meld the commits once you're ready.
Thanks!

**Comment by blairconrad** [comment]
> Also, I'm not sure when the squashing happens. I kept the commits to make the additional addressing of review comments clear. I can meld the commits once you're ready.

Appending commits until the review is done is perfect. I'm gradually trying to spread that knowledge at work since, as you noted, it makes the diffs much easier to read. Some people who send PRs to me regularly squash or at least rebase on every update, and then Bitbucket presents the code as if it's entirely new each time, and I find it a lot more work.

**Review by blairconrad — DISMISSED** [review]
Thanks, @justineclin. Looks good to me.
You indicated an intent to squash, so I'll refrain from merging for now, but I'd've been content to merge as two commits. If you change your mind about squashing though, just tell me.

A note that may influence whether and how you squash: you do have a few different identities going on with your commits:

```
❯ git log -2 --pretty=fuller
commit 0a0c44f19b3b447095acafeb322f47899e5bc8bc (HEAD -> burn-anno-warn, justin/burn-anno-warn)
Author:     Justin Lin <justin.lin@agfa.com>
AuthorDate: Sun Mar 28 01:08:56 2021 -0400
Commit:     Justin Lin <justin.lin@agfa.com>
CommitDate: Sun Mar 28 01:08:56 2021 -0400

    Fix from PR 107 comments - mainly added file id for non-empty warnings.

commit 3297b569ffe6c9fb7d7ce702f50087442ad783cd
Author:     Justin Lin <justin.ec.lin@gmail.com>
AuthorDate: Sun Feb 28 02:34:46 2021 -0500
Commit:     Justin Lin <justin.lin@agfa.com>
CommitDate: Wed Mar 3 21:24:54 2021 -0500

    Added Burned in Annotation Warnings
```

Both commits were committed by you at work, while one was authored by you at home and the other by your at work…
Which is fine, at least from my point of view, but I'm not sure if you'd rather seem like one person, and which person that is. I tend not to commit external work using my work e-mail (partly because I tend not to work on the Open Source projects on my work computer), but it's your choice!

**Comment by justineclin** [comment]
Thanks so much for all the comments, @blairconrad . I didn't know I had two identities (i.e. being batman).

I've always done Dicognito work on the work PC because the set up is nicer - my laptop at home is pushing 10 years (I've given it some RAM as much needed walker/oxygen to hold it over). Hmm. I wonder how the double identity got set up that way. I will look into fixing that, and yes, I'd like to squash, so please wait for me!

I shall be back (Arnold voice).

And thanks again!

**Comment by justineclin** [comment]
I did a squash and pushed the change up here. I like how Bitbucket shows which commits were removed and added - it at least seem to hint at a rebase. I don't see that hint in GitHub - the squashing seemed just like any other commit.
Know of any quick way to see that a commit is just a simple squashing? (Thus avoiding the re-review)

**Comment by blairconrad** [comment]
Hi, @justineclin. You mentioned that the commit looked like any other, but I see this upstairs:

image

showing that you replaced commits. And then I can see

```
❯ git log --graph  --oneline 0a0c44f fe8a6b7  -6
* fe8a6b7 (justin/burn-anno-warn) Added Burned in Annotation Warnings
| * 0a0c44f (HEAD -> burn-anno-warn) Fix from PR 107 comments - mainly added file id for non-empty warnings.
| * 3297b56 Added Burned in Annotation Warnings
|/
*   7bc53bf (origin/master, justin/master, master) Merge pull request #105 from blairconrad/drop-py27-shims
|\
| * 840f0ea Remove shims for Python 2.7
|/
*   a0f974f (tag: 0.12.0) Merge branch 'release/0.12.0'
|\
```

in this case, the base is the same, so the very easy thing for me to do would be to diff the commits:

```
❯ git diff  0a0c44f fe8a6b7
```

which yields

```diff
```

Other than that, I do not know of an easy way to tell that the content was not changed. And it'd be even harder if the base of the branch changed!

Commit looks good to me:

```
commit fe8a6b790e25de7667d2ff2cc49dc4f02d15922e (justin/burn-anno-warn)
Author:     Justin Lin <justineclin@gmail.com>
AuthorDate: Sun Feb 28 02:34:46 2021 -0500
Commit:     Justin Lin <justineclin@gmail.com>
CommitDate: Mon Mar 29 17:30:04 2021 -0400
```

(although really "good to me" would've been anything that was good enough for you!)

**Comment by justineclin** [comment]
Ah, right. So one still has to do a diff - I was hoping to avoid that. Still, it's good to hear your explanation and how you arrived at the conclusion. Thank you very much for the steps!

Looks like I pushed you out of the approved state. Whenever you're ready - no rush.

**Review by blairconrad — APPROVED** [review]

**Pull request merged**